### PR TITLE
don't grow iv list when changing to too complex shape

### PR DESCRIPTION
**Pull request by luke-gru**

### CODE REVIEW

File: variable.c [hunk 1]
```diff
@@ -1440,15 +1441,24 @@ rb_obj_ivar_set(VALUE obj, ID id, VALUE val)
...
+        /* Grow capacity of ivar list. Don't grow capa it if we're about to
+           switch to OBJ_TOO_COMPLEX shape */
+        if (UNLIKELY(!too_complex && shape->next_iv_index >= num_iv)) {
             RUBY_ASSERT(shape->next_iv_index == num_iv);
 
             shape = rb_grow_iv_list(obj);
             RUBY_ASSERT(shape->type == SHAPE_CAPACITY_CHANGE);
         }
 
         rb_shape_t *next_shape = rb_shape_get_next(shape, obj, id);
```
Comment (peterzhu2118) [inline variable.c:1456]: I think the code would be easier to read if we extracted the check for `too_complex` out of `rb_shape_get_next` and instead perform it here instead since this is the only place that calls `rb_shape_get_next` with a `T_OBJECT`.

Comment (luke-gru) [reply variable.c:1456]: Okay I'll do that, thanks for the review!

File: shape.c [hunk 2]
```diff
@@ -581,7 +575,7 @@ rb_shape_too_complex(VALUE self)
 static VALUE
 parse_key(ID key)
 {
-    if (is_instance_id(key)) {
+    if (is_instance_id(key) || is_class_id(key)) {
```
Comment (peterzhu2118) [inline shape.c:578]: Why was this changed? It doesn't seem to be relevant to this PR.

File: shape.c [hunk 3]
```diff
@@ -634,7 +628,7 @@ static VALUE
 rb_shape_edge_name(rb_shape_t * shape)
 {
     if (shape->edge_name) {
-        if (is_instance_id(shape->edge_name)) {
+        if (is_instance_id(shape->edge_name) || is_class_id(shape->edge_name)) {
```
Comment (peterzhu2118) [inline shape.c:631]: Same here, doesn't seem relevant to this PR.

File: shape.c [hunk 4]
```diff
@@ -314,28 +313,23 @@ rb_shape_get_next_iv_shape(rb_shape_t* shape, ID id)
 {
     RUBY_ASSERT(!is_instance_id(id) || RTEST(rb_sym2str(ID2SYM(id))));
     bool dont_care;
-    return get_next_shape_internal(shape, id, SHAPE_IVAR, &dont_care, true);
+    return get_next_shape_internal(shape, id, SHAPE_IVAR, &dont_care);
 }
 
+/**
+ * This function assumes the next shape for the object will not be the "too complex" shape.
```
Comment (peterzhu2118) [inline shape.c:320]: This comment isn't correct I think, it could return the "too complex" shape.

File: variable.c [hunk 5]
```diff
@@ -1440,7 +1441,12 @@ rb_obj_ivar_set(VALUE obj, ID id, VALUE val)
 
         RUBY_ASSERT(!rb_shape_obj_too_complex(obj));
 
-        if (UNLIKELY(shape->next_iv_index >= num_iv)) {
+        VALUE klass = rb_obj_class(obj);
+        bool too_complex = RCLASS_EXT(klass)->variation_count >= SHAPE_MAX_VARIATIONS;
```
Comment (peterzhu2118) [inline variable.c:1445]: Can we move the check `RCLASS_EXT(rb_obj_class(obj))->variation_count >= SHAPE_MAX_VARIATIONS` to a separate function so that we don't have duplicate logic here and `rb_shape_get_next`?

File: variable.c (hunk 1 above)
Comment (peterzhu2118) [reply variable.c:1456]: I see that makes sense.

**Comment by luke-gru** [comment]
I'm closing this because I think it might add more complexity than it's worth. Thanks for your time!
